### ISSUE TITLE
Plans for our selected and active layer model

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#2271]
## 🚀 Feature
The current state is that `selected` is a property that lives on each layer and a `selected` and `deselected` event are emitted. The `active_layer` lives on the viewer and is an attribute.

In #1952 we'll move `active_layer` from the `viewer` on the layerlist, accessible as `viewer.layers.active`. This is now a dynamically calculated, but cached, property. We clear the cache when the selection model changes.

In a follow-up we could drop the `layer.selected` property, and instead have selected live on the layerlist. Selected would now need to be an evented set, something with @tlambert03 has mentioned he can see how to make.

We could then decide do we still need both `layers.active` and `layers.selected` as active would just be the single element of selected if selected has one element, or it would be None (this seems fast enough to recalculated dynamically) so having a `layers.active` attribute still seems good. Might be might want drop the `layers.events.active` and just have the `layers.events.selected` or `layers.selected.events` which emits anytime the selected set changes. If this is the case then we should think again before adding the `active` event to #1952 (instead we should add a selected event).

Once we then make the move to layergroups, being worked on in #2261, the layerslist then becomes a layer group which is also a layer. As a layer it will no longer have a selected attribute (if we removed them as discussed above), but as a layer group, it could now have a selected set attribute. You could still have an `active` attribute but that may now be more dynamically calculated.

Again it looks like we might be left keeping the `layers.active` attribute, but we'd only need a `layers.selected` event.

I think we know enough to decide how we want this to look at the end and then make #1952 look right in order for us to get there. We might though need to added the `EventedSet` and fix the selection model in or before #1952, but I don't think we need to wait for layer groups.

Thoughts @tlambert03?

**Comment by tlambert03** [comment]
couple thoughts:
I continue to like the idea of a selection model that is _independent_ from the item model (this is how Qt does it... not that we need emulate to all Qt things... but in this case I like it).  so a layer group will have the concept of selected items, but via some instance of a `SelectionModel` (this could actually live on `Group` ... not `LayerGroup`.  emphasizing that this is a tight connection between a MutableSequence, and a set of indices in that sequence)

```python
class Group(NestableEventedList, Node):
    def __init__(self):
        self.selection = None  # see setter
	
	@property
	def selection(self):
		return self._selection
	
	@selection.setter
    def selection(self, model=None):
		# optionally accepts a SelectionModel instance
		# allowing two groups to share a selection model
        self._selection = model or SelectionModel()

# where the selection model is an evented set of indices
# (plus one additional attribute for "active")
# and it can tell you whether a given index is selected or active

class SelectionModel(EventedSet):
	def __init__(self):
		self._active = None

	@active.setter
	def active(self, index):
		self._active = index

```

with that setup, you would be able to do something like
```property
layer_selected = layer in layer_group.selection
```

The benefit here is that two `LayerGroup` views (if we want that) could share the same selection model, without each needing to maintain their own independent source of truth (but, they don't have to).

One additional thought:  I feel like `active` and `selected` shouldn't be quite as tightly coupled as they are at the moment.  I don't really like the dynamic inference of which layer is "active" based on the selection.  One should just add/remove layers from the selection set, and set/unset the active layer.

Anyway, while we _could_ set all this up on the existing repo (i.e. without `Group` involved), I honestly feel like it might be a lot of work for no immediate gain.  I think a _lot_ of event connection is going to be trimmed up with layergroups, so my vote would be to hold off on this for now. If you opened this because it's currently blocking you somewhere else, feel free to point that out and we'll discuss

**Comment by sofroniewn** [comment]
> so a layer group will have the concept of selected items, but via some instance of a SelectionModel (this could actually live on Group ... not LayerGroup. emphasizing that this is a tight connection between a MutableSequence, and a set of indices in that sequence)

I like this idea a lot, seems very natural to me

> Anyway, while we could set all this up on the existing repo (i.e. without Group involved), I honestly feel like it might be a lot of work for no immediate gain.

I'm not sure. We might want to use the `EventedList` again and so it might be nice to have selection as something on the list too. For example, we currently use the evented list inside the napari-animation plugin, and more plugin developers might want to use it too.

It might be nice to have both a `SelectableEventedList` and `SelectableGroup` when done that have a similar interface (note we don't need to use those names, but I do like putting selection on the basic objects, and we might want to keep the current objects as they are).

> One additional thought: I feel like active and selected shouldn't be quite as tightly coupled as they are at the moment. I don't really like the dynamic inference of which layer is "active" based on the selection. One should just add/remove layers from the selection set, and set/unset the active layer.

Interesting, so you could in theory have an "active" layer that wasn't selected? I wonder how that gets represented in the UI, is there both "active" and "selection" that get visualized? It hadn't really occurred to me that this would be desired, but I'm very open to this particularly if Qt and other list models do it this way too.

>  If you opened this because it's currently blocking you somewhere else, feel free to point that out and we'll discuss

The question is if we merge #2066 with `viewer.active_layer` still present. I think we can, I'd just add a custom json encoder that for now doesn't try and serialize layers. We probably want that for the layerlist too, so it's probably ok to get #2066 in without this, but that's where I'm focused.

**Comment by tlambert03** [comment]
> >  I honestly feel like it might be a lot of work for no immediate gain.

> I'm not sure. We might want to use the EventedList again and so it might be nice to have selection as something on the list too.

I think we _do_ want all that stuff. my emphasis there was on the "immediate gain" part.  I also like this direction and agree it's reusable, but I'm just very conscious of the fact that the layergroup specifically may very well change a lot in the near future.  so I'm wondering why go through wiring this all up to a layerlist that is about to get removed?  We're _nearly_ there ... 

> so you could in theory have an "active" layer that wasn't selected?

yeah, you could.  I recommend reading [this section](https://doc.qt.io/qt-5/model-view-programming.html#handling-selections-in-item-views) of the qt model/view guide, specifically [this part](https://doc.qt.io/qt-5/model-view-programming.html#current-item-and-selected-items) where they discuss the concepts behind "current" vs "selected"

> The question is if we merge #2066 with viewer.active_layer still present. I think we can, I'd just add a custom json encoder that for now doesn't try and serialize layers. We probably want that for the layerlist too, so it's probably ok to get #2066 in without this, but that's where I'm focused.

Yeah, I don't think having `viewer.active_layer` should be any trouble there.  I like that proposal.  If the only concern is how serialization is going to work, and if we're not using it directly yet, i think we can just "turn it off" for now.

**Comment by sofroniewn** [comment]
Closed by #2441
